internal/logger/logger.go

```diff
@@ -0,0 +1,294 @@
+// set, then the constructor will attempt to source the LogLevel from the environment.
+func New(sink LogSink, maxDocumentLength uint, componentLevels map[Component]Level) *Logger {
+	return &Logger{
+		ComponentLevels: selectComponentLevels(
+			func() map[Component]Level { return componentLevels },
+			getEnvComponentLevels,
+		),
+
+		MaxDocumentLength: selectMaxDocumentLength(
+			func() uint { return maxDocumentLength },
+			getEnvMaxDocumentLength,
+		),
+
+		Sink: selectLogSink(
+			func() LogSink { return sink },
+			getEnvLogSink,
+		),
+
+		jobs: make(chan job, jobBufferSize),
+	}
+
+}
+
+// Close will close the logger and stop the printer goroutine.
+func (logger Logger) Close() {
+	close(logger.jobs)
+}
+
+// Is will return true if the given LogLevel is enabled for the given LogComponent.
+func (logger Logger) Is(level Level, component Component) bool {
+	return logger.ComponentLevels[component] >= level
+}
+
+// TODO: (GODRIVER-2570) add an explanation
+func (logger *Logger) Print(level Level, msg ComponentMessage) {
+	select {
+	case logger.jobs <- job{level, msg}:
+	default:
+		logger.jobs <- job{level, &CommandMessageDropped{}}
+	}
+}
+
+// StartPrintListener will start a goroutine that will listen for log messages and attempt to print them to the
+// configured LogSink.
+func StartPrintListener(logger *Logger) {
+	go func() {
+		for job := range logger.jobs {
+			level := job.level
+			levelInt := int(level)
+
+			msg := job.msg
+
+			// If the level is not enabled for the component, then skip the message.
+			if !logger.Is(level, msg.Component()) {
+				return
+			}
+
+			sink := logger.Sink
+
+			// If the sink is nil, then skip the message.
+			if sink == nil {
+				return
+			}
+
+			keysAndValues, err := formatMessage(msg.Serialize(), logger.MaxDocumentLength)
+			if err != nil {
+				sink.Info(levelInt, "error parsing keys and values from BSON message: %v", err)
+
+			}
+
+			sink.Info(levelInt-DiffToInfo, msg.Message(), keysAndValues...)
+		}
+	}()
+}
+
+func truncate(str string, width uint) string {
+	if len(str) <= int(width) {
+		return str
+	}
+
+	// Truncate the byte slice of the string to the given width.
+	newStr := str[:width]
+
+	// Check if the last byte is at the beginning of a multi-byte character.
+	// If it is, then remove the last byte.
+	if newStr[len(newStr)-1]&0xC0 == 0xC0 {
+		return newStr[:len(newStr)-1]
+	}
+
+	// Check if the last byte is in the middle of a multi-byte character. If it is, then step back until we
+	// find the beginning of the character.
+	if newStr[len(newStr)-1]&0xC0 == 0x80 {
+		for i := len(newStr) - 1; i >= 0; i-- {
+			if newStr[i]&0xC0 == 0xC0 {
+				return newStr[:i]
+			}
+		}
+	}
+
+	return newStr + TruncationSuffix
+}
+
+// TODO: (GODRIVER-2570) remove magic strings from this function. These strings could probably go into internal/const.go
+func formatMessage(keysAndValues []interface{}, commandWidth uint) ([]interface{}, error) {
```

Comment:
This Extended JSON string conversion and truncation logic is only relevant to command logging and should be applied when creating a `CommandMessage` or `CommandSucceededMessage`.